### ISSUE TITLE
macOS, VS + plaformio extension for Greenvawes GAPUINO GAP8 dev. board

### ISSUE THREAD

**Issue by CSEM-LvA** [pioplus/platform-riscv_gap#2]
Hello,

I’ve installed Visual Studio Code 1.3.0, then the PlatformIO extension(Home 2.0.0 - Core 3.6.3).
After installation I imported the example project “190115-162614-gapuino-pulp-os-hello-world”.

The development board is seen by VS (seen in the Devices tab).
The Platforms tab shows RISC-V GAP

The compilation is going fine, but the code cannot be loaded to the target. Here is a copy of the messages in the Terninal tab:

> Executing task in folder 190115-162614-gapuino-pulp-os-hello-world: platformio run --target upload <

Processing gapuino (platform: riscv_gap; board: gapuino; framework: pulp-os)
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https ://docs.platformio.org/page/boards/riscv_gap/gapuino.html
PLATFORM: RISC-V GAP > GAPUINO GAP8 development board
HARDWARE: 250MHz 8MB RAM (64MB Flash)
DEBUG: CURRENT(ftdi) ON-BOARD(ftdi)
Library Dependency Finder -> http ://bit.ly/configure-pio-ldf
LDF MODES: FINDER(chain) COMPATIBILITY(soft)
Collected 0 compatible libraries
Scanning dependencies...
No dependencies
Checking size .pioenvs/gapuino/firmware.elf
Memory Usage -> http ://bit.ly/pio-memory-usage
DATA:    [          ]   0.0% (used 3764 bytes from 8388608 bytes)
PROGRAM: [          ]   0.0% (used 16660 bytes from 67108864 bytes)
Configuring upload protocol...
AVAILABLE: ftdi
CURRENT: upload_protocol = ftdi
Uploading .pioenvs/gapuino/firmware.elf

Traceback (most recent call last):
File "/Users/lva/.platformio/packages/tool-pulp_tools/bin/plpbridge", line 497, in <module>
bridge = db.get_bridge(config=config, verbose=args.verbose, binaries=binaries)
File "/Users/lva/.platformio/packages/tool-pulp_tools/python/bridge/debug_bridge.py", line 45, in get_bridge
return bridge_class(config=config, binaries=binaries, verbose=verbose)
File "/Users/lva/.platformio/packages/tool-pulp_tools/python/bridge/chips/gap.py", line 271, in __init__
super(gap_debug_bridge, self).__init__(config=config, binaries=binaries, verbose=verbose)
File "/Users/lva/.platformio/packages/tool-pulp_tools/python/bridge/default_debug_bridge.py", line 179, in __init__
self.module = ctypes.CDLL(fullpath)
File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/ctypes/__init__.py", line 365, in __init__
self._handle = _dlopen(self._name, mode)
OSError: dlopen(/Users/lva/.platformio/packages/tool-pulp_tools/python/bridge/libpulpdebugbridge.so, 6): Library not loaded: @rpath/libftdi1.2.dylib
Referenced from: /Users/lva/.platformio/packages/tool-pulp_tools/python/bridge/libpulpdebugbridge.so
Reason: image not found

*** [upload] Error 1
=============================================================================== [ERROR] Took 0.94 seconds ===============================================================================
The terminal process terminated with exit code: 1

My system is Mojave 10.14.2 (18C54)

I have found that if I defined the environment variable

export LD_LIBRARY_PATH=~/.platformio//packages/tool-pulp_tools/lib

and that I create a software link to the MacPort libusb

sudo ln -s /opt/local/lib /usr/local/opt/libusb/lib

The in a Terminal I can execute the command:

"/Users/lva/.platformio/penv/bin/python2.7" /Users/lva/.platformio/packages/tool-pulp_tools/bin/plpbridge --debug --verbose=3 --cable=ftdi@digilent --chip=gap --boot-mode=jtag --binary .pioenvs/gapuino/firmware.elf load start

**Comment by ivankravets** [comment]
Could you remove any symbolic links, etc and just `brew install libusb`?

**Comment by CSEM-LvA** [comment]
Thank you for your reply. I applied your instructions but the problem is not solved. Trying to upload the code directly form Visual Studio still produce the error about library not found for ```@rpath/libftdi1.2.dylib``` (see the attached log)
[platformio_error_log.txt](https://github.com/pioplus/platform-riscv_gap/files/2820526/platformio_error_log.txt)

Executing manually the command line in a terminal is working

**Comment by ivankravets** [comment]
Could you try the next?

1. Remove `libusb*` files from `~/.platformio/packages/tool-pulp_tools/lib`
2. Copy here your `libusb*` from `/opt/local/lib`

Does it work without any `export` or `linking`?

**Comment by CSEM-LvA** [comment]
Here are the followed steps:
```
cd ~/.platformio/packages/tool-pulp_tools/lib
mkdir ~/.platformio/packages/backup
mv libusb*  ~/.platformio/packages/backup/
cp -a /opt/local/lib/libusb* .
```
Then VS was launched. Cleaned the project by clicking on the little trashcan in bottom menu bar. Then building the project by clicking the arrow (PlatformIO: Upload). The code is compiling as before, but is not uploaded to the board. Still the same error about a library not loaded as you can see in the attached file.
[verbose_upload_log.txt](https://github.com/pioplus/platform-riscv_gap/files/2832375/verbose_upload_log.txt)

**Comment by ivankravets** [comment]
Do you have compiler on your machine? Could you rebuild a package? Please open system terminal and type:
```
python ~/.platformio/packages/tool-pulp_tools/scripts/build.py
```

**Comment by CSEM-LvA** [comment]
Hello

 

I have Python (2.7 Apple installation) and Xcode. 

 

I also have Python 3.7.2 from python.org.

 

Here is the output of typing the command:

 

python ~/.platformio/packages/tool-pulp_tools/scripts/build.py

Cloning into '/var/folders/g1/_r4yy_054nvc2dcsdt4kzfl40000gp/T/tmpIsvsQO'...

remote: Enumerating objects: 13, done.

remote: Counting objects: 100% (13/13), done.

remote: Compressing objects: 100% (11/11), done.

remote: Total 13 (delta 0), reused 6 (delta 0), pack-reused 0

Unpacking objects: 100% (13/13), done.

Submodule 'gap_flasher' (https://github.com/GreenWaves-Technologies/gap_flasher.git) registered for path 'gap_flasher'

Submodule 'json-tools' (https://github.com/pulp-platform/json-tools) registered for path 'json-tools'

Submodule 'pulp-configs' (https://github.com/pulp-platform/pulp-configs) registered for path 'pulp-configs'

Submodule 'pulp-debug-bridge' (https://github.com/pulp-platform/pulp-debug-bridge) registered for path 'pulp-debug-bridge'

Cloning into '/private/var/folders/g1/_r4yy_054nvc2dcsdt4kzfl40000gp/T/tmpIsvsQO/gap_flasher'...

remote: Enumerating objects: 263, done.        

remote: Total 263 (delta 0), reused 0 (delta 0), pack-reused 263        

Receiving objects: 100% (263/263), 1.80 MiB | 1.40 MiB/s, done.

Resolving deltas: 100% (89/89), done.

Cloning into '/private/var/folders/g1/_r4yy_054nvc2dcsdt4kzfl40000gp/T/tmpIsvsQO/json-tools'...

remote: Enumerating objects: 46, done.        

remote: Counting objects: 100% (46/46), done.        

remote: Compressing objects: 100% (34/34), done.        

remote: Total 194 (delta 20), reused 36 (delta 12), pack-reused 148        

Receiving objects: 100% (194/194), 38.86 KiB | 12.95 MiB/s, done.

Resolving deltas: 100% (85/85), done.

Cloning into '/private/var/folders/g1/_r4yy_054nvc2dcsdt4kzfl40000gp/T/tmpIsvsQO/pulp-configs'...

remote: Enumerating objects: 395, done.        

remote: Counting objects: 100% (395/395), done.        

remote: Compressing objects: 100% (246/246), done.        

remote: Total 5020 (delta 254), reused 267 (delta 148), pack-reused 4625        

Receiving objects: 100% (5020/5020), 1.34 MiB | 1.60 MiB/s, done.

Resolving deltas: 100% (3783/3783), done.

Cloning into '/private/var/folders/g1/_r4yy_054nvc2dcsdt4kzfl40000gp/T/tmpIsvsQO/pulp-debug-bridge'...

remote: Enumerating objects: 246, done.        

remote: Counting objects: 100% (246/246), done.        

remote: Compressing objects: 100% (150/150), done.        

remote: Total 2511 (delta 127), reused 166 (delta 69), pack-reused 2265        

Receiving objects: 100% (2511/2511), 652.16 KiB | 870.00 KiB/s, done.

Resolving deltas: 100% (1505/1505), done.

Submodule path 'gap_flasher': checked out '276515873bdb6868a7debfdb082259dfa5bf6e80'

Submodule path 'json-tools': checked out 'f305a5a3856b7d2a3ea344dad5d6a61a8d2da0e7'

Submodule path 'pulp-configs': checked out '54b9036f121943980f0bd35503a0441c17c3dd46'

Submodule path 'pulp-debug-bridge': checked out '1a9c6c4b34e54cd77e1ca0095d8ff239ee3db2b1'

Traceback (most recent call last):

  File "/Users/lva/.platformio/packages/tool-pulp_tools/scripts/build.py", line 125, in <module>

    build(tmp_dir)

  File "/Users/lva/.platformio/packages/tool-pulp_tools/scripts/build.py", line 64, in build

    subprocess.call(generate, env=env)

  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 522, in call

    return Popen(*popenargs, **kwargs).wait()

  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 710, in __init__

    errread, errwrite)

  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 1335, in _execute_child

    raise child_exception

OSError: [Errno 2] No such file or directory

 

The python version that were used is:

 

python

Python 2.7.10 (default, Aug 17 2018, 19:45:58) 

[GCC 4.2.1 Compatible Apple LLVM 10.0.0 (clang-1000.0.42)] on darwin

 

Thanks for trying to help

 

Best regards

 

Laurent v. Allmen

**Comment by ivankravets** [comment]
Please use Python 2.7. I forgot to mention, you need cmake too. Please install it with `brew install cmake`.

**Comment by CSEM-LvA** [comment]
I just did add cmake in my PATH environment variable as i have seen ii is called by the Python script

 

The installation is now terminating successfuly. I have now to wait tomorrow to try if I can upload code to the Gapuino using Visual Studio as I am not at office rigt now

**Comment by ivankravets** [comment]
Could you pack `~/.platformio/packages/tool-pulp_tools` folder and share somewhere? I would like to try your version of a package which contains binaries built on macOS 10.14. I use 10.13.

**Comment by CSEM-LvA** [comment]
sure

 

i have put a zip archive of my .platformio folder on our OwnCloud server

 

https://owncloud.csem.ch/owncloud/index.php/s/nWH1DKLQdnVtQfP

 

Password : 4P1atf0rmIO

**Comment by ivankravets** [comment]
I see, `libusb` are different on macOS 10.13 & 10.14.

Please try the next:

1. "MacPort libusb" - please remove this libusb by MacPort. I recommend using https://brew.sh/
   `brew install libusb`
2. Remove current `~/.platformio/packages/tool-pulp_tools`, start uploading, PlatformIO will download the latest version of this package which contain `libusb` for macOS 10.13. 
3. Now, please remove `~/.platformio/packages/tool-pulp_tools/lib/libusb-*` files
4. Start uploading.

It should use your Homebrew's `libusb`. Does it work?

**Comment by CSEM-LvA** [comment]
Here are the steps I have followed:

1. sudo port uninstall openocd libftdi1 libusb-compat libusb

2. brew install libusb
   Warning: libusb 1.0.22 is already installed and up-to-date

3. rm -fr ~/.platformio/packages/tool-pulp_tools

4. rm -fr ~/.platformio/packages/tool-pulp_tools/lib/libusb-*

Still not working, I looked in 
ls -al ~/.platformio/packages/tool-pulp_tools/lib/
total 1464
drwxr-xr-x  6 lva  staff     192 Feb 25 07:38 .
drwx------  9 lva  staff     288 Feb 25 07:26 ..
-rw-r--r--  1 lva  staff   60076 Nov  8 12:17 libftdi1.2.4.0.dylib
-rw-r--r--  1 lva  staff   60076 Nov  8 12:17 libftdi1.2.dylib
-rw-r--r--  1 lva  staff   60076 Nov  8 12:17 libftdi1.dylib
-rw-r--r--  1 lva  staff  561768 Nov  8 12:17 libjson.a

And added also from brew:

5. brew install libftdi

Then I could upload and run the code from Visual Code

Thanks for your help
